govcd/edgegateway_lb_service_monitor.go

```diff
@@ -0,0 +1,187 @@
+		return nil, err
+	}
+
+	httpPath, err := eGW.getProxiedEdgeEndpoint("/loadbalancer/config/monitors")
+	if err != nil {
+		return nil, fmt.Errorf("could not get Edge Gateway API endpoint: %s", err)
+	}
+
+	// Anonymous struct to unwrap "monitor response"
+	lbMonitorResponse := &struct {
+		LBMonitors []*types.LBMonitor `xml:"monitor"`
+	}{}
+
+	// This query returns all service monitors as the API does not have filtering options
+	_, err = eGW.client.ExecuteRequest(httpPath, http.MethodGet, "application/xml", "unable to read Load Balancer monitor: %s", nil, lbMonitorResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	// Search for monitor by ID or by Name
+	for _, monitor := range lbMonitorResponse.LBMonitors {
+		// If ID was specified for lookup - look for the same ID
+		if lbMonitorConfig.ID != "" && monitor.ID == lbMonitorConfig.ID {
+			return monitor, nil
+		}
+
+		// If Name was specified for lookup - look for the same Name
+		if lbMonitorConfig.Name != "" && monitor.Name == lbMonitorConfig.Name {
+			// We found it by name. Let's verify if search ID was specified and it matches the lookup object
+			if lbMonitorConfig.ID != "" && monitor.ID != lbMonitorConfig.ID {
+				return nil, fmt.Errorf("load balancer monitor was found by name (%s), but it's ID (%s) does not match specified ID (%s)",
+					monitor.Name, monitor.ID, lbMonitorConfig.ID)
+			}
+			return monitor, nil
+		}
+	}
+
+	return nil, fmt.Errorf("could not find load balancer service monitor (name: %s, ID: %s)",
+		lbMonitorConfig.ID, lbMonitorConfig.Name)
+}
+
+// UpdateLBServiceMonitor
+func (eGW *EdgeGateway) UpdateLBServiceMonitor(lbMonitorConfig *types.LBMonitor) (*types.LBMonitor, error) {
+	if err := validateUpdateLBServiceMonitor(lbMonitorConfig); err != nil {
+		return nil, err
+	}
+
+	// if only name was specified for update, ID must be found, because ID is mandatory for update
+	if lbMonitorConfig.ID == "" {
+		readLBMonitor, err := eGW.ReadLBServiceMonitor(&types.LBMonitor{Name: lbMonitorConfig.Name})
+		if err != nil {
+			return nil, fmt.Errorf("unable to find load balancer monitor by name for update: %s", err)
+		}
+		lbMonitorConfig.ID = readLBMonitor.ID
+	}
+
+	httpPath, err := eGW.getProxiedEdgeEndpoint("/loadbalancer/config/monitors/" + lbMonitorConfig.ID)
+	if err != nil {
+		return nil, fmt.Errorf("could not get Edge Gateway API endpoint: %s", err)
+	}
+
+	// Result should be 204, if not we expect an error of type types.NSXError
+	_, err = eGW.client.ExecuteRequestHTTPCodeOrTypedError(http.StatusNoContent, httpPath, http.MethodPut, "application/xml", "%s", lbMonitorConfig, &types.NSXError{})
+	if err != nil {
+		return nil, fmt.Errorf("error while updating load balancer service monitor : %s", err)
+	}
+
+	readMonitor, err := eGW.ReadLBServiceMonitor(&types.LBMonitor{ID: lbMonitorConfig.ID})
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve monitor with ID (%s) after update: %s", readMonitor.ID, err)
+	}
+	return readMonitor, nil
```

Comment:
This is very good that we return what we created and yet it's not current govcd API behaviour. I thinks it's a point for discussion.

Reply:
Ok. Maybe we can start a discussion even here. The point is that I must return the ID for created object somehow as these functions don't rely solely on names as was the case earlier. My idea is to return whole created object including the ID.

Reply:
In this particular case I am not sure 100% is needed. As when pass parameters for update, you already know what information you updating and details of service monitor.
In general I am saying it is good, only consistency thing under question. @lvirbalas @dataclouder your opinions?

Reply:
I guess we have an agreement on that.

Reply:
yep